dts/bindings/morse/zephyr,morse.yaml

```
@@ -0,0 +1,49 @@
+# Copyright (c) 2024-2025 Freedom Veiculos Eletricos
+# Copyright (c) 2024-2025 OS Systems
+# SPDX-License-Identifier: Apache-2.0
+
+description: Morse Code
+
+compatible: "zephyr,morse"
+
+include:
+  - name: base.yaml
+
+properties:
+  speed:
+    type: int
+    default: 20
+    description: Define the number of Words Per Minute (wps) should be send
+      per minute. This is defined using the standard word "PARIS". In the
+      default case means that "20 wpm" send the word "PARIS" (or precisely
+      "PARIS ") is sent 20 times in a minute.
+
+      In morse the "PARIS " is 50 units long and represented by
+      ".--. .- .-. .. .../" were
+
+        10 dits (.)              10 units
+        4 dahs (-)               12 units
+        9 intra-character spaces  9 units
+        4 inter-character spaces 12 units
+        1 word space (/)          7 units
+
+      In this case, the driver defines the time unit as 60/(50 * Speed). With a
+      default speed of 20 wps the time unit is equal to 0.06s.
+
+  rx_tmr:
+    type: phandle
+    description: The Counter Timer compatible that will be used to deserialize
+      the data into morse code at RX units.
+
+  tx_tmr:
+    type: phandle
+    description: The Counter Timer compatible that will be used to serialize
+      the data into morse code at TX units.
+
+  rx_drv:
+    type: phandle
+    description: The RX time unit driver
+
+  tx_drv:
+    type: phandle
+    description: The TX time unit driver
```

DTS Properties have to use hyphens instead of underscores in the name.
Also I'd like to point to https://github.com/zephyrproject-rtos/zephyr/pull/77650, as counters can use alarms with multiple channels, this would allow to use a single counter device for both TX and RX.

```suggestion
  rx-tmr:
    type: phandle
    description: The Counter Timer compatible that will be used to deserialize
      the data into morse code at RX units.

  tx-tmr:
    type: phandle
    description: The Counter Timer compatible that will be used to serialize
      the data into morse code at TX units.

  rx-drv:
    type: phandle
    description: The RX time unit driver

  tx-drv:
    type: phandle
    description: The TX time unit driver
```

---

> DTS Properties have to use hyphens instead of underscores in the name.

my bad, I forget about this, fixed!

> Also I'd like to point to https://github.com/zephyrproject-rtos/zephyr/pull/77650, as counters can use alarms with multiple channels, this would allow to use a single counter device for both TX and RX.

I'm not sure if it is possible to use the same counter for 2 different things. The channels should work asynchronously in regards to the time base. You need to imagine that you can have a TX on one frequency and an RX on another and the delay between then depends on the round trip time. I don't want to impose restrictions about how this can be implemented otherwise a single timer on a half duplex could be enough.

---

And what if you only have 1 counter instance and need full duplex? I'd assume TX and RX will likely use the same frequency, but you can set the alarm ticks to be different too.

---

Unfortunately it is not possible.

Try imagine that A start to TX. This means that the counter is from 0 to TOP (dot space time). This counter is on the middle and then the RX start to receive. When this counter is on the TOP it will reset and generate an INT to next TX info and you last the RX because time was TOP / 2. The receive needs a counter that is >9 times (dot space time) to generate the interrupt. How we can count the from 0 ~ TOP * 9 ? It is not possible to reach that.

So,you need 2 counter with different TOP values.

> EDIT: native_sim has only 1 counter instance by default

`native_sim` (as it is a sim) should be possible to extend : )

---

Counters also have alarms, can those be used instead?

---

Looking more on it it, I believe it is possible, yes.

---

How I get the amount of time passed between the set alarm and now?

---

#85018